### PR TITLE
Implement initial version of autograd with named tensors

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25604 Implement initial version of autograd with named tensors**

In this initial version:
- autograd ignores all names.
- tensor.grad is unnamed, unless the user manually assigns to it.
- if a grad tensor has any names, perhaps the user was hoping for some
alignment-checking behavior that named tensor offers for other ops. We
raise a warning in this case.

Future: do some more extensive checking to see if this actually works in
all cases.

Test Plan:
- [namedtensor ci]
- Check a warning is raised if a grad tensor has names.
- Check tensor.grad field is unnamed.
- Check that we can perform backward on an op that doesn't explictly
support names in backward. `sigmoid` is one such op.

Differential Revision: [D17171788](https://our.internmc.facebook.com/intern/diff/D17171788)

### CODE REVIEW

File: torch/csrc/autograd/function.h [hunk 1]
```diff
@@ -115,6 +115,12 @@ struct TORCH_API Node : std::enable_shared_from_this<Node> {
     RECORD_FUNCTION(
         this, std::vector<c10::IValue>(inputs.begin(), inputs.end()));
 
+#ifdef BUILD_NAMEDTENSOR
+    // In the first iteration of named tensors, autograd ignores names and
+    // operates on unnamed tensors. In the long term, autograd should
+    // probably operate with names.
+    at::NoNamesGuard no_names_guard;
```
Comment (zou3519) [inline torch/csrc/autograd/function.h]: @ezyang is it guaranteed that all calls to autograd function go through `operator()`? I tried looking through the code and I think this is what happens.

Comment (ezyang) [reply torch/csrc/autograd/function.h]: I think so. If you want to be 100% sure you could make the apply() function virtual private, which will catch at compile time if anyone else naughtily calling apply().

File: torch/csrc/autograd/python_engine.cpp [hunk 2]
```diff
@@ -131,7 +131,17 @@ PyObject *THPEngine_run_backward(THPEngine *self, PyObject *args, PyObject *kwar
...
-      grads.push_back(((THPVariable*)grad)->cdata);
+      const Variable& grad_var = ((THPVariable*)grad)->cdata;
+#ifdef BUILD_NAMEDTENSOR
+      if (grad_var.has_names()) {
+        TORCH_WARN(
+            "Autograd was passed a named grad tensor with dims ", grad_var.names(),
+            ". Autograd does not yet support named tensor semantics, so all names ",
+            "will be ignored. In practice all computed gradients will still be correct "
+            "according to regular tensor semantics.");
+      }
```
Comment (ezyang) [inline torch/csrc/autograd/python_engine.cpp]: In the terminal state, I guess what you should have is the explicit grad does name matching with the output, and then you're off to the races. But this is fine for now :)

**Review by ezyang — APPROVED** [review]
That was easy!

File: torch/csrc/autograd/python_engine.cpp (hunk 2 above)
Comment (zou3519) [reply torch/csrc/autograd/python_engine.cpp]: Agreed. If I have time before the initial release I'll take a stab at doing this.

**Pull request closed without merging**